modules/tensor_mechanics/include/utils/RankTwoScalarTools.h

```diff
@@ -194,46 +334,159 @@ Real hoopStress(const RankTwoTensor & stress,
-                          Point & normalPosition);
+  // Compute the scalar stress component in the direction of the normal vector from the
+  // user-defined axis of rotation.
+  T radial_stress = 0.0;
+  for (unsigned int i = 0; i < 3; ++i)
+    for (unsigned int j = 0; j < 3; ++j)
+      radial_stress += radial_norm(j) * stress(j, i) * radial_norm(i);
+
+  direction = radial_norm;
+
+  return radial_stress;
+}
 
 /*
  * This method calculates the scalar value of the supplied rank-2 tensor in the
  * direction specified by the user.
  */
-Real directionValueTensor(const RankTwoTensor & r2tensor, Point & direction);
+template <typename T>
+T
+directionValueTensor(const RankTwoTensorTempl<T> & r2tensor, Point & direction)
+{
+  T tensor_value_in_direction = 0.0;
+  for (unsigned int i = 0; i < 3; ++i)
+    for (unsigned int j = 0; j < 3; ++j)
+      tensor_value_in_direction += direction(j) * r2tensor(j, i) * direction(i);
+
+  return tensor_value_in_direction;
+}
 
 /*
  * Triaxiality is the ratio of the hydrostatic stress to the von Mises stress.
  */
-Real triaxialityStress(const RankTwoTensor & stress);
+template <typename T>
+T
+triaxialityStress(const RankTwoTensorTempl<T> & stress)
+{
+  return hydrostatic(stress) / vonMisesStress(stress);
+}
 
 /*
  * maxShear is the maximum shear stress defined as the maximum principal
  * stress minus the minimum principal stress.
  */
-Real maxShear(const RankTwoTensor & stress);
-
+template <typename T>
+T
+maxShear(const RankTwoTensorTempl<T> & stress)
+{
+  Point dummy;
+  return (maxPrincipal(stress, dummy) - minPrincipal(stress, dummy)) / 2.;
+}
 /*
  * stressIntensity is defined as two times the maximum shear stress.
  */
-Real stressIntensity(const RankTwoTensor & stress);
+template <typename T>
+T
+stressIntensity(const RankTwoTensorTempl<T> & stress)
+{
+  return 2. * maxShear(stress);
 }
 
+/*
+ * Return scalar quantity of a rank two tensor based on the user specified scalar_type
+ * @param point1 The starting point of the rotation axis for a cylinderical system
+ * @param point2 The end point of the rotation axis
+ * @param curr_point The point corresponding to the stress (pass in & _q_point[_qp])
+ * @param direction The direction vector in which the scalar stress value is calculated
+ * point1 and point2 are required only for the cases of axialStress, hoopStress and radialStress
+ * curr_point is required only for the cases of hoopStress and radialStress
+ * direction is required only for directionValueTensor
+ * for all other cases, these parameters will take the default values
+ */
+template <typename T>
+T
+getQuantity(const RankTwoTensorTempl<T> & tensor,
+            const MooseEnum scalar_type,
+            const Point & point1,
+            const Point & point2,
+            const Point & curr_point,
+            Point & direction)
+{
+  T val = 0.0;
+
+  switch (scalar_type)
+  {
+    case 0:
+      val = vonMisesStress(tensor);
+      break;
```

Comment:
the pattern here should be to return the value directly instead of assigning it. That would also allow you to get rid of teh `break`. Not necessarily in the scope of this PR, but it would be a nice cleanup.

Reply:
good point

Reply:
shoot, I was just changing this...I'll do a new MR